devices/ICub/src/ICub.cpp

```diff
@@ -0,0 +1,388 @@
+/*
+ * Copyright (C) 2018 Istituto Italiano di Tecnologia (IIT)
+ * All rights reserved.
+ *
+ * This software may be modified and distributed under the terms of the
+ * GNU Lesser General Public License v2.1 or any later version.
+ */
+
+#include "ICub.h"
+
+#include <yarp/os/LogStream.h>
+#include <yarp/os/Network.h>
+#include <yarp/os/BufferedPort.h>
+
+#include <map>
+#include <mutex>
+#include <string>
+#include <vector>
+
+using namespace wearable::devices;
+
+const std::string LogPrefix = "ICub :";
+
+class ICub::ICubImpl
+{
+public:
+    class ICubForceTorque6DSensor;
+
+    wearable::TimeStamp timeStamp;
+
+    size_t nSensors;
+    std::vector<std::string> sensorNames;
+
+    bool leftHandFTDataReceived  = false;
+    bool rightHandFTDataReceived = false;
+
+    template<typename T>
+    struct icubDeviceSensors
+    {
+        std::shared_ptr<T> icubSensor;
+        size_t index;
+    };
+
+    std::map<std::string, icubDeviceSensors<ICubForceTorque6DSensor>> ftSensorsMap;
+
+    const WearableName wearableName = "ICub" + wearable::Separator;
+
+    class wrenchPort : public yarp::os::BufferedPort<yarp::os::Bottle> {
+    public:
+        std::vector<double> wrenchValues;
+        bool firstDataReceived = false;
+        std::mutex mtx;
+
+        using yarp::os::BufferedPort<yarp::os::Bottle>::onRead;
+        virtual void onRead(yarp::os::Bottle& wrench) {
```

Comment:
Are you sure you cannot use a callback on `yarp::sig::Vector`? The logic of the callback function would be easier than operating on a `Bottle`.